### DISCUSSION TITLE
Continue parsing on validation-error in list of nested models

### DISCUSSION THREAD

**Discussion by k-o-n-t-o-r** [pydantic/pydantic#2864] [Question]
Hey!
I have the following structure:

```python
from typing import List
from pydantic import BaseModel

class Child(BaseModel):
    test: int

class Parent(BaseModel):
    children: List[Child]

inputObj = {
	"children":  [
		{ "test": 0 },
		{ "test": None },
		{ "test": "aaa" }
	]
}

o = Parent.parse_obj(inputObj)
print(o)
```
As expected, this results in a `ValidationError`.
However, I'd prefer to simply skip children that fail validation instead of aborting the whole process.
In the case above,  `o` would contain a single child with `test: 0`.

I know this is possible using validators by calling `parse_obj` on each child separately.
Does a more ergonomic way to specify this behavior exist?

**Comment by eddsalkield** [comment]
Hi there :smile: 

I don't think there's currently a good way of dynamically controlling data generation to decide what to do when validation fails.  Several approaches have previously been suggested, including [catching the `ValidationError` and parsing its contents](https://github.com/samuelcolvin/pydantic/issues/668#issuecomment-512059268), [interfacing with `validate_model` directly](https://github.com/samuelcolvin/pydantic/issues/1864#issuecomment-679044432), and the addition of a [`check` method](https://github.com/samuelcolvin/pydantic/issues/668#issuecomment-753689197) to separate construction and validation.

I'm also interested in solving this problem, and I've proposed an approach in #2866.  In it, I define a dynamic data generation function `AllSuccessfulOf`, that would allow you to express a solution to your issue as follows:

```python3
o = Parent(children = AllSuccessfulOf(
                   {'test': 0},
                   {'test': None},
                   {'test': 'aaa'}
    ))
```

It would be possible to express whether to log the non-fatal exceptions (like the last two objects failing) with a flag, like so:


```python3
o = Parent(__handle_non_fatal__ = print,
                   children = AllSuccessfulOf(
                       {'test': 0},
                       {'test': None},
                       {'test': 'aaa'},
                       log_non_fatal = True
    ))
```

This would cause non-fatal errors in the `AllSuccessfulOf` to be captured, and handled by the print function.

**Reply by k-o-n-t-o-r** [reply]
Thank you for the insight! In case someone finds this thread in the future, my crude solution looks as follows:

```python

from typing import List
from pydantic import BaseModel, root_validator, Field, ValidationError

class ErrorContainer(object):
    child_errors = list()

def genValidChildren(children, model):
    for c in children:
        try:
            yield model.parse_obj(c)
        except ValidationError as e:
            ErrorContainer.child_errors.append(e)

class KeepValidatingOnError(BaseModel):
    @root_validator(pre=True)
    def traverse_sources(cls, values):
        for (name, field) in cls.__fields__.items():
            keep_validating_on_error = field.field_info.extra.get(
                "keep_validating_on_error", False
            )
            if keep_validating_on_error:
                values[name] = list(genValidChildren(values[name], field.type_))

        return values



# Example: --------------------------------------------------------------


class ExampleChildModel(BaseModel):
    value: int

class ExampleParentModel(KeepValidatingOnError):
    idents: List[ExampleChildModel] = Field(keep_validating_on_error=True)

o = {"idents": [{"value": 1}, {"value": 2}, {"value": "aa"}, {"value": None}]}


parsed = ExampleParentModel.parse_obj(o)
# idents=[ExampleChildModel(value=1), ExampleChildModel(value=2)]


child_errors = ErrorContainer.child_errors
#	[
# 		ValidationError(model='ExampleChildModel',
# 			errors=[{
# 			'loc': ('value',),
# 			'msg': 'value is not a valid integer,
# 			'type': 'type_error.integer'
# 			}]
# 		) ...
# ]

```
